contracts/mmr/MMRVerification.sol

```diff
@@ -8,124 +8,140 @@ pragma solidity ^0.7.0;
  *
  *      The index of this MMR implementation starts from 1 not 0.
  */
-library MMRVerification {
+contract MMRVerification {
+
+    struct MountainData {
+        uint256 position;
+        bytes32 hash;
+        uint256 height;
+    }
+
+    mapping(uint256 => MountainData) public queue;
 
     /**
-     * @dev It returns true when the given params verifies that the given value exists in the MMR.
+     * @dev Verify an MMR inclusion proof for a leaf at a given index.
      */
-    function inclusionProof(
+    function verifyInclusionProof(
         bytes32 root,
-        uint256 width,
-        uint256 index,
-        bytes32 value,
-        bytes32[] memory peaks,
-        bytes32[] memory siblings
+        bytes32 leafNodeHash,
+        uint256 leafIndex,
+        uint256 leafCount,
+        bytes32[] memory proofItems
```

Comment:
Why do we need `leafIndex` and `leafCount` here? Should not the `leafNodeHash`, `root` and `proofItems` be enough to prove inclusion?

Reply:
Excluding edge cases, our proof items' contents are organized by index from left-to-right:
1. From the left-most index to our leaf's peak index: individual peak hashes
2. From our leaf's peak index to our leaf's index: sibling hashes of our leaf
3. From our leaf's index to the right-most index: all the right peaks pre-hashed together into one hash

We need the `leafIndex` in order to properly climb the branch from leaf -> peak, as the hashing order of siblings changes depending on if the node is the left or right sibling.

The Nervos Network MMR library actually [passes the contents](https://github.com/nervosnetwork/merkle-mountain-range/blob/master/src/mmr.rs#L238) of all the leaves and the MMR size into the `calculate_root` function, from which the content of individual leaves and `leafCount` is used in the `calculate_peaks_hashes` function.

We're using `leafCount` to get the peak indexes. From our leaf's index and the peak indexes, we can calculate our leaf's peak index. When climbing the branch we track the current parent index and once it reaches our expected peak index we return the hash we've built by traversing the branch. It may be possible to iterate over the proof items array every time we calculate a parent hash to see if it matches any of the proof items, at which point we know we've reached the peak. I'll need to think more about the potential implications of this approach.

Reply:
Makes sense to me, thanks for elaborating. 

I did indeed also end upusing count/width in the positioned merkle proofs, since an odd tree implies you need to skip hashing. 

I'm fairly certain you could derive the count/width of the tree/MMR from the proof size though, but it's fine for now if we skip this as a potential optimization.